compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/ResolveUtils.kt

```diff
@@ -247,9 +247,35 @@ private fun BodyResolveComponents.typeFromSymbol(symbol: AbstractFirBasedSymbol<
     }
 }
 
-fun BodyResolveComponents.transformQualifiedAccessUsingSmartcastInfo(qualifiedAccessExpression: FirQualifiedAccessExpression): FirQualifiedAccessExpression {
+fun BodyResolveComponents.transformQualifiedAccessUsingSmartcastInfo(
+    qualifiedAccessExpression: FirQualifiedAccessExpression
+): FirQualifiedAccessExpression {
     val typesFromSmartCast = dataFlowAnalyzer.getTypeUsingSmartcastInfo(qualifiedAccessExpression) ?: return qualifiedAccessExpression
     val originalType = qualifiedAccessExpression.resultType.coneType
+    // For example, if (x == null) { ... },
+    //   we don't want to smartcast to Nothing?, but we want to record the nullability to its own kind of node.
+    // TODO: should we differentiate x == null v.s. x is Nothing?
+    if (typesFromSmartCast.any { it.isNullableNothing }) {
+        val typesFromSmartcastWithoutNullableNothing =
+            typesFromSmartCast.filterTo(mutableListOf()) { !it.isNullableNothing }.also {
```

Comment:
Here, I only discarded `Nothing?`, but an alternative I tried was:
```kt
val nullableTypesFromSmartcastWithoutNullableNothing =
    typesFromSmartcast.mapNotNullTo(mutableListOf()) {
        if (it.isNullableNothing) null else it.withNullability(...NULLABLE)
    }...
```
i.e., making everything else _nullable_ (since it's cast to literally `null`).

But, encountered non-sense cases, e.g.,
```kt
if (x != null) {
  ...
  if (x == null) argShouldNotNull(<!>x<!>)
}
```
So, using the old info (from `x != null`) without `Nothing?` doesn't introduce any change, so I ended up with that.